tuning/tuned_model.py

```diff
@@ -0,0 +1,226 @@
+"""Interface for loading and running trained causal LMs. In the future,
+these capabilities will be unified with the sft_trainer's tuning capabilities.
+"""
+
+import datasets
+import os
+from typing import Optional, Union
+
+from peft import AutoPeftModelForCausalLM
+from peft.utils.other import fsdp_auto_wrap_policy
+import torch
+from transformers import AutoModelForCausalLM, AutoTokenizer, LlamaTokenizer, LlamaTokenizerFast, GPTNeoXTokenizerFast
+from transformers.utils import logging
+from transformers import TrainerCallback
+from trl import SFTTrainer, DataCollatorForCompletionOnlyLM
+
+from tuning.aim_loader import get_aimstack_callback
+from tuning.config import configs, peft_config
+from tuning.data import tokenizer_data_utils
+from tuning.utils import AdapterConfigPatcher
+from tuning.utils.config_utils import get_hf_peft_config
+from tuning.utils.data_type_utils import get_torch_dtype
+
+class PeftSavingCallback(TrainerCallback):
+    def on_save(self, args, state, control, **kwargs):
+        checkpoint_path = os.path.join(args.output_dir, f"checkpoint-{state.global_step}")
+        kwargs["model"].save_pretrained(checkpoint_path)
+
+        if "pytorch_model.bin" in os.listdir(checkpoint_path):
+            os.remove(os.path.join(checkpoint_path, "pytorch_model.bin"))
+
+
+class TunedCausalLM:
+    def __init__(self, model, tokenizer, device):
+        self.peft_model = model
+        self.tokenizer = tokenizer
+        self.device = device
+
+    @classmethod
+    def load(cls, checkpoint_path: str, base_model_name_or_path: str=None) -> "TunedCausalLM":
+        """Loads an instance of this model.
+
+        Args:
+            checkpoint_path: str
+                Checkpoint model to be loaded, which is a directory containing an
+                adapter_config.json.
+            base_model_name_or_path: str [Default: None]
+                Override for the base model to be used.
+
+        By default, the paths for the base model and tokenizer are contained within the adapter
+        config of the tuned model. Note that in this context, a path may refer to a model to be
+        downloaded from HF hub, or a local path on disk, the latter of which we must be careful
+        with when using a model that was written on a different device.
+
+        Returns:
+            TunedCausalLM
+                An instance of this class on which we can run inference.
+        """
+        overrides = {"base_model_name_or_path": base_model_name_or_path} if base_model_name_or_path is not None else {}
```

Comment:
Currently the adapter config patcher is an internal detail, and we only use it to override the base model, since it does not seem like we can do this from the automodel API, but it's a helpful capability to have